File: haskell/fizzbuzz.hs
-- fizzbuzz :: Integer -> String
-- fizzbuzz n
--     | mod n 3 == 0 && mod n 5 == 0 = "FizzBuzz"
--     | mod n 3 == 0 = "Fizz"
--     | mod n 5 == 0 = "Buzz"
--     | otherwise = show n
-- main = do putStrLn $ unlines (map fizzbuzz [1..100])


module Main where
data Rule = Rule Int String
myrules :: [Rule]
myrules = 
    [ Rule 3 "Fizz"
    , Rule 5 "Buzz"
    , Rule 7 "Bazz"
    ]
fizzbuzz :: [Rule] -> Int -> String
fizzbuzz rules n
  = case [s | Rule d s <- rules, d `divides` n] of
      [] -> show n
      strings -> concat strings
  where d `divides` m = m `rem` d == 0

main :: IO ()
main = putStr . unlines $ map (fizzbuzz myrules) [1..100]

File: ocaml/fizzbuzz.ml
(* let fizzbuzz i = *)
(*   match i mod 3, i mod 5 with *)
(*     0, 0 -> "FizzBuzz" *)
(*   | 0, _ -> "Fizz" *)
(*   | _, 0 -> "Buzz" *)
(*   | _    -> string_of_int i *)
let (=>) condition f x = if condition then f x else x
let append str a = a^str
let fizzbuzz i = 
    "" |> (i mod 3 = 0 => append "Fizz")
       |> (i mod 5 = 0 => append "Buzz")
       |> (function "" -> string_of_int i | s -> s)
let _ =
  for i = 1 to 100 do print_endline (fizzbuzz i) done

